### ISSUE TITLE
[BUG] Returned values from quadtree_on_point vary between runs

### ISSUE THREAD

**Issue by drabastomek** [rapidsai/cuspatial#318]
**Describe the bug**
`quadtree_on_point` returns vary between runs and eventually (if rerun enough times) led to a bad memory access.

**Steps/Code to reproduce bug**
I used the example from the docs

```
[0]: np.random.seed(2)
points = cudf.DataFrame({
        "x": cudf.Series(np.random.normal(size=120)) * 500,
        "y": cudf.Series(np.random.normal(size=120)) * 500,
    })

max_depth = 3
min_size = 50
min_x, min_y, max_x, max_y = (points["x"].min(),
                                  points["y"].min(),
                                  points["x"].max(),
                                  points["y"].max())
scale = max(max_x - min_x, max_y - min_y) // (1 << max_depth)
print(
        "min_size:   " + str(min_size) + "\n"
        "num_points: " + str(len(points)) + "\n"
        "min_x:      " + str(min_x) + "\n"
        "max_x:      " + str(max_x) + "\n"
        "min_y:      " + str(min_y) + "\n"
        "max_y:      " + str(max_y) + "\n"
        "scale:      " + str(scale) + "\n"
    )


[1]: key_to_point, quadtree = cuspatial.quadtree_on_points(
        points["x"],
        points["y"],
        min_x,
        max_x,
        min_y,
        max_y,
        scale, max_depth, min_size
    )

print(quadtree)
print(key_to_point)
```

When cell `[1]` is run multiple times, every now and then (randomly) I would get odd results (larger quadtree). Upon ~30 reruns it started complaining about bad memory access. I think there might be a memory-leak somewhere.

Normal run produces the following:

```
   key  level  is_quad  length  offset
0    0      0    False      33       0
1    1      0    False      37      33
2    2      0    False      18      70
3    3      0    False      31      88
4    6      0    False       1     119
0       63
1       20
2       66
3      117
4       19
      ... 
115     31
116     69
117    116
118     98
119     24
Length: 120, dtype: uint32
```
An abnormal one, the following

```
    key  level  is_quad  length  offset
0     0      0    False      33       0
1     1      0    False      37      33
2     2      0    False      18      70
3     3      0    False      31      97
4     6      0    False       1     159
5    38      2    False       1      88
6    39      2    False       5      89
7    41      2    False       1      94
8    45      2    False       1      95
9    47      2    False       1      96
10   48      2    False      10     128
11   49      2    False       6     138
12   50      2    False       1     144
13   51      2    False       1     145
14   52      2    False       4     146
15   53      2    False       3     150
16   54      2    False       1     153
17   55      2    False       1     154
18   56      2    False       2     155
19   57      2    False       1     157
20   58      2    False       1     158
21  106      2    False       1     160
0       63
1       20
2       66
3      117
4       19
      ... 
115     31
116     69
117    116
118     98
119     24
Length: 120, dtype: uint32
```

**Expected behavior**
Neither of the results match the documentation -- I would expect them to match. Also, between reruns the results should remain stable since I'm not changing the input data.

**Environment details (please complete the following information):**
 - Environment location: Docker
 - Method of cuSpatial install: Docker
   - Stable git pull from rapidsai/rapidsai, v 0.16, CUDA 11, Python 3.7.8

**Comment by zhangjianting** [comment]
@drabastomek Thanks for your bug report. I will look into these issues and get back to you ASAP. --Jianting

**Comment by drabastomek** [comment]
Thanks @zhangjianting! Let me know if I can be of any help!

**Comment by zhangjianting** [comment]
@drabastomek What is the GPU model that you are using? cudf 0.15 works fine on my Titan V machine but cudf 0.16 failed quite some of its unit tests on my RTX Titan machine. I want to exclude the possibility due to cudf. 
If you can test on cudf/cuspatial 0.15 and let me know whether the code works, it would be great. Your code seems to work fine for me on my Titan V with cudf/cuspatial 0.15.

**Comment by zhangjianting** [comment]
It seems that C++ and python have different results. 
For the [python test code (with randomly generated data)](http://geoteci.engr.ccny.cuny.edu/cuspatial/point_quadtree_debug.cu), Python output has 5 1st level quadrants and all are leaf nodes, which does not make sense. 

 key  level  is_quad  length  offset
0    0      0    False      27       0
1    1      0    False      25      27
2    2      0    False      33      [52]
3    3      0    False      34      85
4    8      0    False       1     119

However, [the C++ test code]((http://geoteci.engr.ccny.cuny.edu/cuspatial/point_quadtree_debug.cu)  output , output 4 nodes, which seems to be correct. 
         0          0          0         33          0
         1          0          0         32         33
         2          0          0         25         65
         3          0          0         30         90
@trxcllnt when you have a chance can you take a look?

**Comment by zhangjianting** [comment]
@drabastomek @trxcllnt It seems that there is an issue in [computing scale](https://github.com/rapidsai/cuspatial/blob/a06630547dbcdd710a590e936a8ba0bb589efd2d/python/cuspatial/cuspatial/core/indexing.py#L92)
Python integer division (//) operator rounds down (i.e., floor in C++) which makes computed scale lower than it needs to be. This would make points closer to boundary having a row/column ID larger than (2^max_depth-1) and the Z-order code is computed incorrectly. Problems then follows. 
I would suggest change it to `scale = max(max_x - min_x, max_y - min_y)  ((1 << max_depth)-2) `. Since scale needs to be a floating number, there is no need to use integer division. 
Also, we need to change [scale calculation](https://github.com/rapidsai/cuspatial/blob/a06630547dbcdd710a590e936a8ba0bb589efd2d/python/cuspatial/cuspatial/core/indexing.py#L158) as well to correctly detect the situations that scale is too small. (change from +2 to -2 in the line). 
@drabastomek  It seems that the minor bug does not affect the correctness of the output if you compute/set scale correctly in your code. Can you give it a try again? Thanks.

**Comment by zhangjianting** [comment]
@drabastomek I ran a [copy of your code outside of a notebook ](http://geoteci.engr.ccny.cuny.edu/cuspatial/bug318_quadtree.py) and the results are stable.  Can you run the code and see how it behaves. The expected output of the quadtree is: 
 
key  level  is_quad  length  offset
0    0      0     True       4       4
1    1      0    False      34      72
2    2      0    False      10     106
3    3      0    False       4     116
4    0      1    False       8       0
5    1      1    False      20       8
6    2      1    False       9      28
7    3      1    False      35      37

**Comment by drabastomek** [comment]
I got the same output as you did. Still, when I rerun the cell in a notebook it looks like there's some memory leak somewhere.

**Comment by kkraus14** [comment]
> I got the same output as you did; when I run the cell once I get this result in a notebook as well. Still, when I rerun the cell in a notebook it looks like there's some memory leak somewhere.

When running a cell IPython will automatically keep the output of the cell available as a variable (and subsequently keep a reference to the object which prevents cleaning it up).

Are you actually using `print(quadtree)` or just outputting `quadtree` in the cell? If the cell is `[1]`, then the output of the cell would be stored in `_1`.

**Comment by drabastomek** [comment]
@kkraus14 I am printing using `print(quadtree)`.

**Comment by zhangjianting** [comment]
> I got the same output as you did; when I run the cell once I get this result in a notebook as well. Still, when I rerun the cell in a notebook it looks like there's some memory leak somewhere.
I can confirm that inconsistency happens when I run cuspatial.quadtree_on_points in a loop. By adding "quadtree=None" and "key_to_point=None" after the call, the inconsistency is gone.  
As such, can you put  to put "quadtree=None" and "key_to_point=None" BEFORE you call cuspatial.quadtree_on_points? It might make a difference when you rerun the cell. The idea is to release variables of previous results to avoid potential CPU/GPU memory problems. 
I am not a python (notebook) experts, hopefully someone can explain better.

**Comment by drabastomek** [comment]
I can confirm that setting the `quadtree` and `key_to_point` variables before assigning them removes the inconsistency. This tells me that the variable needs to be assigned and looks like the memory needs to be allocated/cleared before running the method. 

I think the problem is in the communication between C++ and Python layers: right now the function returns as follows 

```
cpdef quadtree_on_points(...):
    ...
    return (
        Column.from_unique_ptr(move(result.first)),
        Table.from_unique_ptr(
            move(result.second),
            column_names=["key", "level", "is_quad", "length", "offset"]
        )
    )
```

Perhaps there some issue with Cython here? Or maybe we should be initializing temp `Column` and `Table` variables before returning them? I do not know Cython that well (learning currently) so this is just a hunch...

**Comment by shwina** [comment]
The Cython looks fine. Is there any chance this is related to https://github.com/jupyter/notebook/issues/3713? Also, could we get a minimal reproducer of the problem perhaps?

**Comment by drabastomek** [comment]
Potentially that might be the culprit. But I did not notice such behavior with other parts of RAPIDS. The minimal example to reproduce this is in the first message.

**Comment by kkraus14** [comment]
It sounds somewhat like we're expecting memory to be zeroed on initialization which isn't always the case. So when the function allocates memory for its output column(s) there's situations where it's getting random data instead of the expected zeroes.

@zhangjianting is it possible this is happening from the C++ side? I.E. did we switch to `device_uvector` recently which doesn't initialize the memory versus `device_vector`?

**Comment by zhangjianting** [comment]
@kkraus14 good point. We did use device_uvector on purpose for efficiency. I think we made sure the values of device_uvector elements were filled after the function completes but I will double check when I get a chance.  
I am still trying to understand why setting python variables to None solved the inconsistency problem. If a python variable (e.g., quadtree) was assigned a returned (valid) value after the first call to cuspatial.quadtree_on_points, what is the effect of setting the variable to None? Seems that it behaves like a destructor in C++ to release memory but I am not sure. 
On the other hand, for p= cuspatial.quadtree_on_point(...), if p obtains a valid value for the first call, if the line is executed again without p=None, then what will happen to the memory that was previously associated with p? I would imagine a memory leak in this case, but p should still point to valid memory chunk allocated during the second call. However, in this case, the new p should point to new memory chunk and I am not sure uvector could be a reason for the issue. 
Is there a documentation on how cudf python memory model works? 
@shwina Can you comment on this as well?

**Comment by shwina** [comment]
There's this: https://docs.python.org/3/c-api/memory.html#memory-management

When cuDF objects get garbage collected, the destructor for any C++ objects underneath them will be called and the device memory released.

However, according to the document above, and from our observations as well, there is **no** guarantee about _when_ objects get GC'd -- even objects that are "unreachable". 

I'm curious what the behavior is when encapsulating with a function. Anecdotally, I've seen the following behaviour on the command-line/Jupyter:

```
>>> a = Foo()
>>> a = 2  # previous `a` may or may not be GCd
```

```
>>> def bar():
...         a = Foo()
...         a = 2
...
>>> bar()  # GC occurs as expected
```

**Comment by zhangjianting** [comment]
@trxcllnt @kkraus14 @harrism 
It seems that Keith was right. [A uninitialized device_uvector was used in thrust::scatter](https://github.com/rapidsai/cuspatial/blob/branch-0.16/cpp/src/indexing/construction/detail/phase_2.cuh#L195)
Adding the following line between declaration and thrust::scatter seems solve the issue: 
thrust::uninitialized_fill(rmm::exec_policy(stream)->on(stream),parent_pos.begin(),parent_pos.end(),0);
The [test code](https://github.com/zhangjianting/cuspatial_benchmark_nyctaxi/blob/main/cpp/tests/indexing/point_quadtree_toy_debug.cu) output consistent results for each of the 10 loop steps.  A copy of the output is in the same dir.

**Comment by zhangjianting** [comment]
> There's this: https://docs.python.org/3/c-api/memory.html#memory-management
> 
> When cuDF objects get garbage collected, the destructor for any C++ objects underneath them will be called and the device memory released.
> 
> However, according to the document [here](https://docs.python.org/3/reference/datamodel.html), and from our observations as well, there is **no** guarantee about _when_ objects get GC'd -- even objects that are "unreachable".
> 
> I'm curious what the behavior is when encapsulating with a function. Anecdotally, I've seen the following behaviour on the command-line/Jupyter:
> 
> ```
> >>> a = Foo()
> >>> a = 2  # previous `a` may or may not be GCd
> ```
> 
> ```
> >>> def bar():
> ...     a = Foo()
> ...     a = 2
> ...
> >>> bar()  # GC occurs as expected
> ```

Thanks.

**Comment by harrism** [comment]
@zhangjianting would be great if you can put the fix in a PR. Scatter is an easy one to miss, because most times when you run an algorithm with a device_uvector as output the algorithm touches every element in the output range, but scatter is different. Your solution is correct, and is still better than `thrust::device_vector` because `device_vector` not only calls `uninitialized_fill`, but it also calls `cudaDeviceSynchronize()`.

**Comment by zhangjianting** [comment]
> @zhangjianting would be great if you can put the fix in a PR. Scatter is an easy one to miss, because most times when you run an algorithm with a device_uvector as output the algorithm touches every element in the output range, but scatter is different. Your solution is correct, and is still better than `thrust::device_vector` because `device_vector` not only calls `uninitialized_fill`, but it also calls `cudaDeviceSynchronize()`.

@harrism I will be glad to create a PR. I had this kind of mistakes several times before. 
I still do not fully understood what has happened exactly. For the [toy test code](https://github.com/zhangjianting/cuspatial_benchmark_nyctaxi/blob/fb69ba9632ed871ec7eed1faa7efbca44f7539fa/cpp/tests/indexing/point_quadtree_toy_debug.cu#L84), quadtree_pair is declared before the loop and its value is changed in each of the l0 loop steps. To me,  quadtree_pair should point to different new memory blocks after each loop step and memory allocated in its previous steps simply got lost (since there is no deallocation).  As the toy problem has a really small memory footprint, it should use new and uninitialized memory chunks for returns. The only explanation to me is that local variables (e.g., [parent_pos ](https://github.com/rapidsai/cuspatial/blob/branch-0.16/cpp/src/indexing/construction/detail/phase_2.cuh#L195) ) were allocated to memory chunks that were previously used but returned afterwards.  Please correct me if I am wrong. 
It would be great if someone can provide an verify the fix works before requesting a merge. This would a little more work than code review. @trxcllnt

**Comment by harrism** [comment]
The way to verify the fix works is to create a test that fails without the fix (ideally a gtest, but a pytest can be added as well). Then, if the fix and test are both correct, the test should pass.

**Comment by kkraus14** [comment]
Fixed by #320
